### PR TITLE
Add new indexes and flag to model_run_results

### PR DESCRIPTION
**Pull request by eKatia**
To be able to analyse models per execution, we need to have additional indexes and flags.

- `model_invocation_index` - would allow us to compare different models executions, e.g. comparison performance per model after having done some tuning. 
- `is_the_last_invocation_of_the_day` - helps to identify the latest invocation that happened to the models on that day.
- `is_the_first_invocation_of_the_day` - helps to compare overall models performance over days by selecting the first invocation per day ( here is the assumption that the first invocation of the day is the main models run)

### CODE REVIEW

File: models/edr/run_results/model_run_results.sql [hunk 1]
```diff
@@ -36,6 +36,10 @@ SELECT
     models.path,
     models.original_path,
     models.owner,
-    models.alias
+    models.alias,
+    ROW_NUMBER() OVER (PARTITION BY unique_id ORDER BY generated_at DESC, execute_started_at DESC, execute_completed_at DESC ) AS model_invocation_index,
+    CASE WHEN DENSE_RANK() OVER (ORDER BY generated_at DESC) = 1 THEN TRUE ELSE FALSE END AS is_latest_invocation, 
+    CASE WHEN DENSE_RANK() OVER (PARTITION BY generated_at::date ORDER BY generated_at ASC) = 1 THEN TRUE ELSE FALSE END AS is_the_first_invocation_of_the_day
```
Comment (eKatia) [inline models/edr/run_results/model_run_results.sql:42]:
```suggestion
    ROW_NUMBER() OVER (PARTITION BY run_results.unique_id ORDER BY run_results.generated_at DESC, run_results.execute_started_at DESC, run_results.execute_completed_at DESC ) AS model_invocation_index,
    CASE WHEN DENSE_RANK() OVER (ORDER BY run_results.generated_at DESC) = 1 THEN TRUE ELSE FALSE END AS is_latest_invocation, 
    CASE WHEN DENSE_RANK() OVER (PARTITION BY run_results.generated_at::date ORDER BY run_results.generated_at ASC) = 1 THEN TRUE ELSE FALSE END AS is_the_first_invocation_of_the_day
```

File: models/edr/run_results/model_run_results.sql [hunk 2]
```diff
@@ -36,6 +36,10 @@ SELECT
     models.path,
     models.original_path,
     models.owner,
-    models.alias
+    models.alias,
+    ROW_NUMBER() OVER (PARTITION BY run_results.unique_id ORDER BY run_results.generated_at DESC, run_results.execute_started_at DESC, run_results.execute_completed_at DESC ) AS model_invocation_index,
+    CASE WHEN DENSE_RANK() OVER (ORDER BY run_results.generated_at DESC) = 1 THEN TRUE ELSE FALSE END AS is_latest_invocation, 
+    CASE WHEN DENSE_RANK() OVER (PARTITION BY run_results.generated_at::date ORDER BY run_results.generated_at ASC) = 1 THEN TRUE ELSE FALSE END AS is_the_first_invocation_of_the_day
```
Comment (IDoneShaveIt) [inline models/edr/run_results/model_run_results.sql:42]: `::date` is not a cross db command.
You can use `elementary.time_trunc` macro instead to make sure it works across all of our supported adapters 🙂 
```suggestion
    CASE WHEN DENSE_RANK() OVER (PARTITION BY {{ elementary.time_trunc('day', 'run_results.generated_at') }} ORDER BY run_results.generated_at ASC) = 1 THEN TRUE ELSE FALSE END AS is_the_first_invocation_of_the_day
```

File: models/edr/run_results/model_run_results.sql [hunk 3]
```diff
@@ -36,6 +36,10 @@ SELECT
...
+
```
Comment (IDoneShaveIt) [inline models/edr/run_results/model_run_results.sql:43]: Notice that another model named `alerts_dbt_models` is based on `model_run_results` 👀 
Currently it does a union between `model_run_results` and `snapshot_run_results`, which is failing due to mismatched column numbers between the 2 models.

You can either:
1. Select the specific columns in the union at the model `alerts_dbt_models` (My preferred solution)
2.  Add those columns to the `snapshot_run_results` model

**Review by IDoneShaveIt — CHANGES_REQUESTED** [review]
Hey @eKatia! Thank you for the contribution! 😃 
The PR looks good overall, but I left some small comments 🙂

Can you share with us how are you going to use this changes once they are merged?
Maybe we can help and implement those cases in our CLI and report?
We would love to learn from your use cases and improve our product!

File: models/edr/run_results/model_run_results.sql [hunk 4]
```diff
@@ -36,6 +36,10 @@ SELECT
     models.path,
     models.original_path,
     models.owner,
-    models.alias
+    models.alias,
+    ROW_NUMBER() OVER (PARTITION BY run_results.unique_id ORDER BY run_results.generated_at DESC, run_results.execute_started_at DESC, run_results.execute_completed_at DESC ) AS model_invocation_index,
```
[the remark this answers is not in the document]
Comment (eKatia) [reply models/edr/run_results/model_run_results.sql:40]: For me it's important to have it in DESC order, that way I can always select the last invocation of a model in a BI tool. E.g. i am doing some performance improvement work, and I want to compare the last 2 executions -> so I always filter by model_invocation_index in (1,2)

File: models/edr/run_results/model_run_results.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (eKatia) [reply models/edr/run_results/model_run_results.sql:42]: Yes, that's right! I guess adding the last invocation of the day would also make sense and be more meaningful than global latest.

File: models/edr/run_results/model_run_results.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (IDoneShaveIt) [reply models/edr/run_results/model_run_results.sql:40]: Got it!
So maybe in that case we should give this column a more indicative name 🙂 
Something like `model_invocation_reverse_index`?

File: models/edr/run_results/model_run_results.sql (hunk 3 above)
Comment (eKatia) [reply models/edr/run_results/model_run_results.sql:43]: Done :)

File: models/edr/run_results/model_run_results.sql (hunk 4 above)
Comment (eKatia) [inline models/edr/run_results/model_run_results.sql:40]:
```suggestion
    ROW_NUMBER() OVER (PARTITION BY run_results.unique_id ORDER BY run_results.generated_at DESC) AS model_invocation_reverse_index,
```

[the remark this answers is not in the document]
Comment (eKatia) [reply models/edr/run_results/model_run_results.sql:40]: sure! Check out the suggestion below

File: models/edr/alerts/alerts_dbt_models.sql [hunk 5]
```diff
@@ -6,9 +6,40 @@
 }}
 
 with error_models as (
-    select * from {{ ref('model_run_results') }}
+  
+    select model_execution_id,
```
Comment (IDoneShaveIt) [inline models/edr/alerts/alerts_dbt_models.sql:10]: The original select was `select *` and it contains more columns 🙂 
To make sure we don't break anything that might be relaying on that model, lets add the rest of the columns:
```
model_execution_id,
unique_id,
invocation_id,
name,
generated_at,
status,
full_refresh,
message,
execution_time,
execute_started_at,
execute_completed_at,
compile_started_at,
compile_completed_at,
compiled_code,
database_name,
schema_name,
materialization,
tags,
package_name,
path,
original_path,
owner,
alias
```

File: models/edr/alerts/alerts_dbt_models.sql [hunk 6]
```diff
@@ -6,9 +6,40 @@
...
+           tags,
+           alias,
+           status,
+           full_refresh 
+    from {{ ref('model_run_results') }}
+  
     union all
-    select * from {{ ref('snapshot_run_results') }}
+  
+    select model_execution_id,
```
Comment (IDoneShaveIt) [inline models/edr/alerts/alerts_dbt_models.sql:28]: See previous comment 🙂

**Review by IDoneShaveIt — COMMENTED** [review]
Looking good!
Another really small comments, and we are ready to go!

File: models/edr/alerts/alerts_dbt_models.sql (hunk 5 above)
Comment (eKatia) [reply models/edr/alerts/alerts_dbt_models.sql]: Updated

**Review by IDoneShaveIt — APPROVED** [review]
Looking good!
Thanks again for the contribution!

File: models/edr/run_results/model_run_results.sql [hunk 7]
```diff
@@ -36,6 +36,10 @@ SELECT
     models.path,
     models.original_path,
     models.owner,
-    models.alias
+    models.alias,
+    ROW_NUMBER() OVER (PARTITION BY run_results.unique_id ORDER BY run_results.generated_at DESC) AS model_invocation_reverse_index,
+    CASE WHEN DENSE_RANK() OVER (ORDER BY run_results.generated_at DESC) = 1 THEN TRUE ELSE FALSE END AS is_latest_invocation, 
+    CASE WHEN DENSE_RANK() OVER (PARTITION BY {{ elementary.time_trunc('day', 'run_results.generated_at') }} ORDER BY run_results.generated_at ASC) = 1 THEN TRUE ELSE FALSE END AS is_the_first_invocation_of_the_day
```
Comment (eKatia) [inline models/edr/run_results/model_run_results.sql:42]:
```suggestion
    CASE WHEN DENSE_RANK() OVER (PARTITION BY {{ elementary.time_trunc('day', 'run_results.generated_at') }} ORDER BY run_results.generated_at ASC) = 1 THEN TRUE ELSE FALSE END AS is_the_first_invocation_of_the_day,
    CASE WHEN DENSE_RANK() OVER (PARTITION BY {{ elementary.time_trunc('day', 'run_results.generated_at') }} ORDER BY run_results.generated_at DESC) = 1 THEN TRUE ELSE FALSE END AS is_the_last_invocation_of_the_day
```

**Pull request merged**